File: downloadblob.py
# -*- coding: utf-8 -*-
"""
Created on Thu Apr  6 11:37:08 2023

@author: ASUS
"""

import os
from azure.storage.blob import BlobServiceClient
from azure.core.exceptions import ResourceExistsError

storage_connection_string = '<storage connection string>'
blob_service_client = BlobServiceClient.from_connection_string(storage_connection_string)

# Create a container
container_id = '<container id>'
container_client = blob_service_client.create_container(container_id)
# container_client = blob_service_client.get_container_client(container_id)
print(container_client)

# List containers
all_containers = blob_service_client.list_containers(include_metadata=True)
for container in all_containers:
    print(container)


# Upload a blob to a container
blob_root_directory = 'dbt'
working_dir = os.getcwd()
file_directory = os.walk(working_dir + '/dbt')
for folder in file_directory:
    for file in folder[-1]:
        
        try:
            file_path = os.path.join(folder[0], file)      
            blob_path = '{0}{1}'.format(
                blob_root_directory,
                file_path.replace(r"C:\Users\Ready_(Azure Storage) Blob Management/dbt", '')
            )
            blob_obj = blob_service_client.get_blob_client(container=container_id, blob=blob_path)
            with open(file_path, mode='rb') as file_data:
                blob_obj.upload_blob(file_data)
        except ResourceExistsError as e:
            print('Blob (file object) {0} already exists.'.format(file))
            continue
        except Exception as e:            
            raise Exception(e)


# List blobs (file objects) in a given container
blobs = container_client.list_blobs()
# blobs = container_client.list_blobs(name_starts_with='dbt')
for blob in blobs:
    print(blob['name'])
    print(blob['container'])
    print(blob['snapshot'])
    print(blob['version_id'])
    print(blob['is_current_version'])
    print(blob['blob_type'])
    print(blob['blob_tier'])
    print(blob['metadata'])
    print(blob['creation_time'])
    print(blob['last_modified'])
    print(blob['last_accessed_on'])
    print(blob['size'])
    print(blob['deleted'])
    print(blob['deleted_time'])
    print(blob['tags'])    


# Download a blob
file_object_path = 'dbt/2. Build dbt projects/1. Build your DAG/Exposures dbt Developer Hub.pdf'
file_downloaded = os.path.join(working_dir, 'Exposures dbt Developer Hub.pdf')

with open(file_downloaded, mode='wb') as download_file:
    download_file.write(container_client.download_blob(file_object_path).readall())

"""
# Delete a blob (subfolder in this example)
blobs = container_client.list_blobs()
for blob in blobs:
    if blob.name.startswith('dbt/List of commands/'):
        container_client.delete_blob(blob.name)
        print('Blob {0} deleted'.format(blob.name))

# Delete a container
container_client.delete_container()
"""

File: download_blob.cs
using System;
using System.IO;
using Azure.Storage.Blobs;
using Azure.Storage.Blobs.Models;
using Azure.Core.Exceptions;

class Program
{
    static void Main(string[] args)
    {
        // Set the connection string and create a BlobServiceClient object
        string connectionString = "<storage connection string>";
        BlobServiceClient blobServiceClient = new BlobServiceClient(connectionString);

        // Create a container
        string containerName = "<container id>";
        BlobContainerClient containerClient;
        try
        {
            containerClient = blobServiceClient.CreateBlobContainer(containerName);
        }
        catch (RequestFailedException ex) when (ex.ErrorCode == BlobErrorCode.ContainerAlreadyExists)
        {
            containerClient = blobServiceClient.GetBlobContainerClient(containerName);
        }
        Console.WriteLine(containerClient.Uri);

        // List containers
        foreach (BlobContainerItem container in blobServiceClient.GetBlobContainers())
        {
            Console.WriteLine(container.Name);
            Console.WriteLine(container.Properties.Metadata);
        }

        // Upload a blob to a container
        string blobRootDirectory = "dbt";
        string workingDir = Directory.GetCurrentDirectory();
        foreach (string folder in Directory.GetDirectories(Path.Combine(workingDir, "dbt"), "*", SearchOption.AllDirectories))
        {
            foreach (string file in Directory.GetFiles(folder))
            {
                try
                {
                    string filePath = Path.Combine(folder, file);
                    string blobPath = $"{blobRootDirectory}{filePath.Replace(@"C:\Users\Ready_(Azure Storage) Blob Management\dbt", string.Empty)}";
                    BlobClient blobClient = containerClient.GetBlobClient(blobPath);
                    using (FileStream fileStream = new FileStream(filePath, FileMode.Open))
                    {
                        blobClient.Upload(fileStream, true);
                    }
                }
                catch (RequestFailedException ex) when (ex.ErrorCode == BlobErrorCode.BlobAlreadyExists)
                {
                    Console.WriteLine($"Blob (file object) {file} already exists.");
                }
                catch (Exception ex)
                {
                    throw new Exception(ex.Message);
                }
            }
        }

        // List blobs (file objects) in a given container
        foreach (BlobItem blob in containerClient.GetBlobs())
        {
            BlobProperties properties = blob.Properties;
            Console.WriteLine(blob.Name);
            Console.WriteLine(blob.ContainerName);
            Console.WriteLine(properties.Snapshot);
            Console.WriteLine(properties.VersionId);
            Console.WriteLine(properties.IsCurrentVersion);
            Console.WriteLine(properties.BlobType);
            Console.WriteLine(properties.AccessTier);
            Console.WriteLine(properties.Metadata);
            Console.WriteLine(properties.CreatedOn);
            Console.WriteLine(properties.LastModified);
            Console.WriteLine(properties.LastAccessed);
            Console.WriteLine(properties.ContentLength);
            Console.WriteLine(properties.DeletedTime);
            Console.WriteLine(properties.Tags);
        }

        // Download a blob
        string fileObjectPath = "dbt/2. Build dbt projects/1. Build your DAG/Exposures dbt Developer Hub.pdf";
        string fileDownloaded = Path.Combine(workingDir, "Exposures dbt Developer Hub.pdf");
        BlobDownloadInfo blobDownloadInfo = containerClient.GetBlobClient(fileObjectPath).Download();
        using (FileStream downloadFileStream = new FileStream(fileDownloaded, FileMode.Create))
        {
            blobDownloadInfo.Content.CopyTo(downloadFileStream);
        }
    }
}
